### PR TITLE
Make SMO instruction take data ptr as an argument

### LINKED ISSUE #396: Make smo use rb for msg data ptr and rc for data len
_Originally posted by @xgreenx in https://github.com/FuelLabs/fuel-vm/pull/373#discussion_r1139692913_

### PR DESCRIPTION
**Pull request by Dentosal**
Closes #396

### CODE REVIEW

File: fuel-vm/src/tests/blockchain.rs [hunk 1]
```diff
@@ -1322,8 +1322,8 @@ fn smo_instruction_works() {
         #[rustfmt::skip]
         let script = vec![
```
Comment (xgreenx) [inline fuel-vm/src/tests/blockchain.rs:1323]: Could you also fix the "TODO" below, please?=) It requires checking of the data output=)

Comment (Dentosal) [reply fuel-vm/src/tests/blockchain.rs:1326]: Done. Also fixed some issues I found after that.

File: fuel-vm/src/interpreter/blockchain.rs [hunk 2]
```diff
@@ -699,14 +692,16 @@ impl<Tx> MessageOutputCtx<'_, Tx> {
...
-            self.call_abi_len as usize,
-            memory_constraint,
-        )?;
+
+        if self.msg_data_len > MEM_MAX_ACCESS_SIZE {
+            return Err(RuntimeError::Recoverable(PanicReason::MemoryOverflow));
+        }
+
+        if self.msg_data_len > self.max_message_data_length {
+            return Err(RuntimeError::Recoverable(PanicReason::MemoryOverflow));
```
Comment (xgreenx) [inline fuel-vm/src/interpreter/blockchain.rs:701]: It may be better to return some named error here.

File: fuel-vm/src/tests/blockchain.rs [hunk 3]
```diff
@@ -1319,13 +1319,21 @@ fn smo_instruction_works() {
         let secret = SecretKey::random(rng);
         let sender = rng.gen();
 
+        // Two bytes of random data to send as the message
+        let msg_data = [rng.gen::<u8>(), rng.gen::<u8>()];
+
         #[rustfmt::skip]
         let script = vec![
+            op::movi(0x10, 2),                          // data buffer allocation size
```
Comment (xgreenx) [inline fuel-vm/src/tests/blockchain.rs:1327]: Maybe with `script_with_data_offset!` we may do it more readable=D

Comment (Dentosal) [reply fuel-vm/src/tests/blockchain.rs:1327]: This is not script_data, instead I just allocate heap space for the bytes. Do you think I should rather use script_data for this?

File: fuel-vm/src/interpreter/blockchain.rs (hunk 2 above)
Comment (Dentosal) [reply fuel-vm/src/interpreter/blockchain.rs]: Done, `MessageDataTooLong` is the new error type.

**Review by xgreenx — APPROVED** [review]
It's up to decide regarding `script_with_data_offset!`. The change already is good, thank you!=)

File: fuel-vm/src/tests/blockchain.rs (hunk 3 above)
Comment (Voxelot) [reply fuel-vm/src/tests/blockchain.rs:1327]: We probably shouldn't use script_with_data_offset anymore now that we have `op::gtf` to find the start of the script data.

**Review by Voxelot — APPROVED** [review]

**Pull request merged**